CoreScriptsRoot/Modules/VR/NotificationHub.lua

```diff
@@ -0,0 +1,730 @@
+
+				local notificationsEnd = #group.notifications
+				if notificationsEnd > 0 then
+					local notificationsStart = math.max(1, notificationsEnd - MAX_NOTIFICATIONS_SHOWN + 1)
+					for i = 1, notificationsEnd do
+						local notification = group.notifications[i]
+						if i >= notificationsStart then
+							notification.frame.Visible = true
+							notification.frame.Position = UDim2.new(0, notificationOffset, 0, notificationOffset)
+							notificationOffset = notificationOffset + NOTIFICATION_PADDING_Y
+
+							local subpanel = notificationsPanel:SetSubpanelDepth(notification.frame, notificationDepth)
+							notificationDepth = notificationDepth + NOTIFICATION_DEPTH_OFFSET
+						else
+							notification.frame.Visible = false
+						end
+					end
+
+					if activeGroup == group then
+						notificationsStart = math.max(1, notificationsEnd - MAX_DETAILS_SHOWN + 1)
+						local detailY = 0
+						local fraction = 1 / MAX_DETAILS_SHOWN
+						for i = 1, notificationsEnd do
+							local notification = group.notifications[i]
+							if i >= notificationsStart then
+								notification.detailsFrame.Visible = true
+								notification.detailsFrame.Position = UDim2.new(0, DETAILS_PADDING, detailY, DETAILS_PADDING)
+								notification.detailsFrame.Size = UDim2.new(1, -DETAILS_PADDING * 2, fraction, -DETAILS_PADDING * 2)
+
+								detailY = detailY + fraction
+							else
+								notification.detailsFrame.Visible = false
+							end
+						end
+					end
+				end
+			end
+		end
+	end
+end
+
+local NotificationGroup = {} 
+do
+	local NotificationGroup_mt = { __index = NotificationGroup }
+	function NotificationGroup.new(key, title, order) 
+		local self = setmetatable({}, NotificationGroup_mt)
+
+		self.key = key
+		self.title = title
+		self.order = order
+		self.notifications = {}
+		self.notificationsDirty = false
+		self.frame = Util:Create "Frame" {
+			Parent = notificationsWindow.content,
+			BackgroundTransparency = 1
+		}
+		self.detailsFrame = Util:Create "Frame" {
+			Parent = nil,
+			BackgroundTransparency = 1,
+
+			Position = UDim2.new(0, 0, 0, 0),
+			Size = UDim2.new(1, 0, 1, 0)
+		}
+
+		notificationsGroups[key] = self
+		table.insert(notificationsGroupsList, self)
+
+		return self
+	end
+
+	function NotificationGroup:Deactivate()
+		self.detailsFrame.Parent = nil
+		for i, v in pairs(self.detailsFrame:GetChildren()) do
+			detailsPanel:RemoveSubpanel(v)
+		end
+	end
+
+	function NotificationGroup:SwitchTo()
+		detailsWindow:SetTitle(self.title)
+		for i, v in pairs(notificationsGroups) do
+			if v ~= self then
+				v:Deactivate()
+			end
+		end
+		self.detailsFrame.Parent = detailsWindow.content
+
+		activeGroup = self
+		self.notificationsDirty = true
+	end
+
+	function NotificationGroup:BringNotificationToFront(notification)
+		if activeGroup ~= self then
+			self:SwitchTo()
+		end
+
+		for i, v in ipairs(self.notifications) do
+			if v == notification then
+				if i == #self.notifications then
```

Comment:
it feels like we can factor this if block out of the loop. Before we do the for loop check if the top notification is the one we are looking for